[Sema] Hide underscored names from typo-correction

adamshin opened the pull request:
Identifiers with a leading underscore are now hidden from typo-correction unless the typed name also begins with an underscore.

Resolves [SR-9083](https://bugs.swift.org/browse/SR-9083).

adamshin committed:
[Sema] Hide underscored names from typo-correction

Identifiers with a leading underscore are now hidden from typo-correction unless the typed name also begins with an underscore.

slavapestov commented:
We already have something similar in code completion -- should we share the logic between the two instead of adding new rules?

adamshin commented:
I was looking into the code completion side of things as well, but wasn't sure where that logic was located. Are you saying there's existing logic around leading-underscore names that could be reused here?

rjmccall reviewed (approved):
LGTM.

adamshin commented:
Anything else needed here?

rjmccall commented:
@swift-ci Please test.

rjmccall commented:
That's definitely your failure to fix.

slavapestov commented:
@adamshin It looks like code completion calls Decl::isPrivateStdlibDecl(), which has a number of checks in addition to leading underscore. Can you try using that instead?

jrose-apple commented:
Sorry, I was the one who wrote up the bug this way. We *don't* want to use `isPrivateStdlibDecl` for this, because (1) we want to opt out of it if the user actually writes an underscore, and (2) we want this to apply to all APIs, not just stdlib ones.

jrose-apple commented:
@swift-ci Please test

adamshin commented:
Looks like my changes broke a test (Parse/recovery.swift). I'm looking into that now.

adamshin committed:
[Sema] Fix failing typo-correction test

The unresolved identifier 'esp' used to have '_exp' as a typo-correction option. That's no longer the case, and the type checker now singles out 'test' as the unique correction.

rjmccall commented:
@swift-ci Please test.

jrose-apple commented:
Failure looks unrelated; let's try again.

@swift-ci Please test Linux

jrose-apple commented:
Thanks, Adam!
